src/dune_pkg/opam_repo.ml

```diff
@@ -34,33 +34,77 @@ module Paths = struct
   let opam_file package = Path.Local.relative (package_dir package) "opam"
 end
 
+module Source = struct
+  type committish = Commit of string
+
+  type t =
+    { url : string
+    ; commit : committish option
+    }
+
+  let of_opam_url ({ OpamUrl.hash; _ } as opam_url) =
+    (* OpamUrl doesn't distinguish between branches and commits and neither
+       does git *)
+    let commit = Option.map hash ~f:(fun hash -> Commit hash) in
+    let url = OpamUrl.base_url opam_url in
+    { commit; url }
+  ;;
+
+  let to_string { url; commit = _ } = url
+
+  let to_dyn { url; commit } =
+    let commit_to_dyn (Commit c) = Dyn.variant "Commit" [ Dyn.string c ] in
+    Dyn.record [ "url", Dyn.string url; "commit", Dyn.option commit_to_dyn commit ]
+  ;;
+
+  let equal { url; commit } t =
+    let committish_equal (Commit c) (Commit c') = String.equal c c' in
+    String.equal url t.url && Option.equal committish_equal commit t.commit
+  ;;
+
+  let encode { url; commit = _ } =
+    (* CR-someday Leonidas-from-XIV: encode commit and remove repo id? *)
+    let open Dune_sexp.Encoder in
+    record_fields [ field "url" string url ]
+  ;;
+
+  let decode =
+    let open Dune_sexp.Decoder in
+    fields
+    @@ let+ url = field "url" string in
+       { url; commit = None }
+  ;;
+end
+
 module Serializable = struct
   type t =
     { repo_id : Repository_id.t option
-    ; source : string
+    ; source : Source.t
     }
 
   let equal { repo_id; source } t =
-    Option.equal Repository_id.equal repo_id t.repo_id && String.equal source t.source
+    Option.equal Repository_id.equal repo_id t.repo_id && Source.equal source t.source
   ;;
 
   let to_dyn { repo_id; source } =
     let open Dyn in
     variant
       "opam_repo_serializable"
-      [ option Repository_id.to_dyn repo_id; string source ]
+      [ option Repository_id.to_dyn repo_id; Source.to_dyn source ]
   ;;
 
   let encode { repo_id; source } =
     let open Encoder in
     record_fields
-      [ field "source" string source; field_o "repo_id" Repository_id.encode repo_id ]
+      [ field_l "source" Fun.id (Source.encode source)
```

Comment:
I think this ends up being problematic in the situation where the user specifies a branch, a tag, or a shortened commit hash. We should really record full git object hash in all of these situations.